enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextProviderImpl.java

```diff
@@ -337,25 +324,100 @@ public void registerFileListing( NeoStoreFileListing fileListing )
+            for ( WritableFulltext writableFulltext : writableNodeIndices[newSide].values() )
+            {
+                writableFulltext.awaitNoReaders();
+                writableFulltext.drop();
+                writableFulltext.open();
+                populations.add( applier.populateNodes( writableFulltext, db ) );
+            }
+            for ( WritableFulltext writableFulltext : writableRelationshipIndices[newSide].values() )
+            {
+                writableFulltext.awaitNoReaders();
+                writableFulltext.drop();
+                writableFulltext.open();
+                populations.add( applier.populateRelationships( writableFulltext, db ) );
+            }
+
+            for ( AsyncFulltextIndexOperation population : populations )
+            {
+                population.awaitCompletion();
             }
-            catch ( ExecutionException e )
+            side = newSide;
+            synchronized ( this )
             {
-                throw new IOException( "Unable to prepare index for snapshot." );
+                notifyAll();
             }
-            index.flush();
-            ResourceIterator<File> snapshot = index.snapshot();
-            snapshots.add( snapshot );
-            files.addAll( getSnapshotFilesMetadata( snapshot ) );
+        }
+        catch ( ExecutionException | InterruptedException e )
+        {
+            throw new IOException( "Unable to flip fulltext indexes", e );
+        }
+        finally
+        {
+            configurationLock.writeLock().unlock();
+        }
+    }
+
+    private Resource snapshotStoreFiles( Collection<StoreFileMetadata> files ) throws IOException
+    {
+        final Collection<ResourceIterator<File>> snapshots = new ArrayList<>();
+        Lock lock = configurationLock.readLock();
+        lock.lock();
+        try
+        {
+            for ( WritableFulltext index : Iterables.concat( writableNodeIndices[ASIDE].values(), writableNodeIndices[BSIDE].values(),
+                    writableRelationshipIndices[ASIDE].values(), writableRelationshipIndices[BSIDE].values() ) )
+            {
+                // Save the last committed transaction, then drain the update queue to make sure that we have applied _at least_ the commits the config claims.
+                index.saveConfiguration( transactionIdStore.getLastCommittedTransactionId() );
+                try
+                {
+                    applier.writeBarrier().awaitCompletion();
+                }
+                catch ( ExecutionException e )
+                {
+                    throw new IOException( "Unable to prepare index for snapshot." );
+                }
+                index.flush();
+                ResourceIterator<File> snapshot = index.snapshot();
+                snapshots.add( snapshot );
+                files.addAll( getSnapshotFilesMetadata( snapshot ) );
+            }
+        }
+        finally
+        {
+            lock.unlock();
         }
```

Comment:
Don't we need to keep the configuration read lock for the duration of the snapshotting? What would otherwise happen if we drop the index while reading the snapshot?